GPS: Parsers and Unit Tests

christopherkinyua opened the pull request:
## Related Issues

Closes #16 

## Summary

This PR is the first implementation step for integrating the GNSS Receiver with the system. This includes receiving the ASCII response from the receiver, parsing the response, and passing the parsed responses to their appropriate structs for BESTXYZA and TIMEA log function response.

## Changes Made

The changes made to realize this were as follows:
- Modified `uart_handler` to receive the GNSS response in interrupt mode. 
- Created a FreeRTOS thread task that will handle the receiving and parsing of the GNSS receiver response
- Created files `gps.c` and `gps_types.h` containing the functions, data structs, and enum definitions.
- Added the `unit_test_gps.c` file, which contains all the unit tests related to the GNSS receiver.
- Modified `LOG_message` function to handle a larger buffer for printing to UART
- Added a function `GEN_int64_to_str` that handles converting int64_t values into strings

## Testing Instructions
I verified the accuracy of the GPS functions by running the `run_all_unit_tests` telecommand and ensured that all my unit tests passed. I have left my `LOG_message` function calls in to show the error that is being invoked. This will be removed during the review to declutter the unit test call function with all the prints to UART. 


## Special Notes for Your Reviewer(s)

There are some `TODO` tags within the code based on my current thoughts on the implementation I have done. Let me know if they are justified and we can discuss how to implement them and if there are other considerations I am missing.

During the Hexagon Office Tour, some of the engineers questioned my implementation method of using ASCII instead of binary. They were in favor of using binary mode. @robotoshi had a more in-depth conversation with the engineers than I did, he could inform us what he was talking about. Furthermore, they mentioned that they had an open-source library called EDIE  for ASCII and binary decoding of the GNSS receiver. 

The link is here: https://novatel.com/support/bulletins-alerts-release-notices/hexagon-novatel-announces-version-3-00-of-its-encode-decode-interface-engine

DeflateAwning commented on firmware/Core/Src/rtos_tasks/rtos_tasks.c:
```diff
@@ -163,3 +165,49 @@ void TASK_execute_telecommands(void *argument) {
+
+		// Parsing the gps data
+		gps_response_header gps_header;
+		const u_int8_t gps_header_result = parse_gps_header(latest_gps_response,&gps_header);
```
What's this `u_int8_t` ?

christopherkinyua replied:
Typo, my bad.

DeflateAwning commented on firmware/Core/Src/rtos_tasks/rtos_tasks.c:
```diff
@@ -163,3 +165,49 @@ void TASK_execute_telecommands(void *argument) {
+		// How do I handle the incomplete data? Could drop the data
+		if(gps_header_result != 0){
+			// Failed to parse
+			break;
```
Breaking this loop ends the task, forever (until the next reboot). I wouldn't do that.

christopherkinyua replied:
Sure thing, I replaced it with a `return` instead for now.

DeflateAwning commented on firmware/Core/Src/rtos_tasks/rtos_tasks.c:
```diff
@@ -163,3 +165,49 @@ void TASK_execute_telecommands(void *argument) {
+		osDelay(200);
+
+		// Checking if there is data in the GPS buffer
+		if(UART_gps_buffer_write_idx > 0) {
```
Isn't this going to potentially try to read the GPS data as it's still being sent? Other subsystems use a timeout to determine when it's done sending.

christopherkinyua replied:
So like I delay the system and then check if like new data has come in? That is my first thought on this. 


Ill try move forward with what I said above until you say otherwise.

DeflateAwning commented on firmware/Core/Src/transforms/arrays.c:
```diff
@@ -63,6 +63,51 @@ void GEN_uint64_to_str(uint64_t value, char *buffer) {
+        return;
+    }
+
+    char temp[32];  // Temporary array to hold digits
```
You can remove the comment if you name the variable appropriately.

christopherkinyua replied:
Sure thing

DeflateAwning commented on firmware/Core/Src/transforms/arrays.c:
```diff
@@ -63,6 +63,51 @@ void GEN_uint64_to_str(uint64_t value, char *buffer) {
+
+    char temp[32];  // Temporary array to hold digits
+    int index = 0;
+    int is_negative = 0;  // Flag to track if the number is negative
```
What about `const uint8_t is_negative = (value < 0);`

christopherkinyua replied:
This makes sense.

DeflateAwning commented on firmware/Core/Src/uart_handler/uart_handler.c:
```diff
@@ -77,6 +106,7 @@ void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart) {
 void UART_init_uart_handlers(void) {
     // enable the UART interrupt
     HAL_UART_Receive_IT(&hlpuart1, (uint8_t*) &UART_telecommand_buffer_last_rx_byte, 1);
+    HAL_UART_Receive_IT(UART_gps_port_handle, (uint8_t*) &UART_gps_buffer_last_rx_byte, 1);
```
I think we actually want to enable and disable the GPS interrupt separately. Given that I've seen it spam constant 0's if it gets in that mode, we want to default to it being turned off, and only turn it on when we're ready to record data.

If it gets in the spamming-zeros mode, then the interrupt will be constantly triggered, and the satellite won't be able to do anything else. 

Let's remove this from here, and let's wrap the `HAL_UART_Receive_IT` call within the Interrupt Service Routine (ISR) in an `if (GPS_uart_interrupt_enabled) {}`. Then, let's have a function called `GPS_set_uart_interrupt_state`, which triggers `HAL_UART_Receive_IT`, and sets the `GPS_uart_interrupt_enabled` flag variable.

christopherkinyua replied:
Yeah, i think ive done an implementation of what you have requested. Let me know if it makes sense.

DeflateAwning reviewed (changes requested):
This is looking really awesome! 

A lot of convention changes.

One important, easy-to-implement change with the way we enable interrupts.

christopherkinyua commented on firmware/.stm32env:
```diff
@@ -0,0 +1,7 @@
+# environment variable file used by stm32-for-vscode and the STM32Make.make makefile
+# Other environment variables can be added here. If wanting to use the generated makefile in CI/CD context please
```
Done

christopherkinyua commented on firmware/Core/Inc/gps/gps.h:
```diff
@@ -0,0 +1,17 @@
+
+#include "gps/gps_types.h"
+
+uint32_t crc32_value(uint8_t i);
```
Done. I'll keep note for future functions

christopherkinyua commented on firmware/Core/Src/gps/gps.c:
```diff
@@ -0,0 +1,945 @@
+/// @param status_str The status string to parse.
+/// @param status Pointer to GPS_reference_time_status_t where the status will be stored.
+/// @return Returns 0 on success, 1 if the status string is unrecognized.
+uint8_t assign_gps_time_status(const char *status_str, GPS_reference_time_status_t *status) {
```
Done

christopherkinyua commented on firmware/Core/Src/gps/gps.c:
```diff
@@ -0,0 +1,945 @@
+/// @brief Assigns a string value based on the provided GPS time status.
+/// @param status GPS_reference_time_status_t value.
+/// @return Returns the assigned string value for the GPS time status.
+const char* get_gps_time_status_string(GPS_reference_time_status_t status) {
```
Done

christopherkinyua commented on firmware/Core/Src/gps/gps.c:
```diff
@@ -0,0 +1,945 @@
+    char message_buffer[256];
+    snprintf(
+        message_buffer, sizeof(message_buffer),
+        "{\"Log Name\":\"%s\",\"Time Status\":\"%s\"}\n",
```
Done

christopherkinyua commented on firmware/Core/Inc/gps/gps_types.h:
```diff
@@ -0,0 +1,173 @@
+/// All reported receiver times are subject to a qualifying time status
+/// Refer to page 51-53 of the OEM7 Commands and Logs Reference Manual
+typedef enum {
+    GPS_UNKNOWN = 20,
```
Done. I have also made changes to the position type enum and modified the prefix as : `GPS_SOL_STATUS_SOL_COMPUTED` instead of `GPS_SOL_COMPUTED`

christopherkinyua commented on firmware/Core/Inc/gps/gps_types.h:
```diff
@@ -0,0 +1,173 @@
+    GPS_APPROXIMATE = 60,
+    GPS_COARSEADJUSTING = 80,
+    GPS_COARSE = 100,
+    GPS_COARSESTEERING = 120,
```
Would you prefer I leave it as is? I am not picky about either option, as long as it is readable.

As of now, I have left it as is but let me know if you would prefer if I changed it.

christopherkinyua commented on firmware/Core/Src/gps/gps.c:
```diff
@@ -0,0 +1,945 @@
+}
+
+
+/// @brief Assigns a GPS time status based on the provided string.
```
Done

christopherkinyua commented on firmware/Core/Src/gps/gps.c:
```diff
@@ -0,0 +1,945 @@
+
+
+/// @brief Parse the received GPS header into a struct
+/// @param data_received - The string obtained from the buffer that is to be parsed into the gps_response_header struct
```
I just want to get some clarification, like should we be checking if the passed string is null terminated?

christopherkinyua commented:
@DeflateAwning I've addressed all the PR comments except the one asking for some clarification. Let me know what you think and I can make some updates so you can continue with the PR review

DeflateAwning commented:
Changes required:
* Timeout in receiving from GPS (refer to Telecommand RTOS and/or EPS `send_cmd_get_response` function)
* `continue` in RTOS task on error instead of returning
* Add 1 telecommand called: `TCMDEXEC_gps_set_enabled(0 or 1)` -> For now, just call `GPS_set_uart_interrupt_state()`. Also, add a TODO to call the relevant EPS command to enable/disable power to the GPS.

After that, let's do some testing with this with the actual GPS! This looks really great!

DeflateAwning commented:
Just merged the EPS firmware; you'll have to merge main into this branch, or rebase this branch to main. Sorry about that.

christopherkinyua commented:
@DeflateAwning I have addressed the comments and added code sections based on the discussion on Thursday. I have a question regarding the power channel from the EPS that will be used for the GPS. From the documentation in the OEM7 Installation Operation Manual Page 29, the GPS receiver operates with 3.3VDC +-5% with less than 100mV ripple. 

I have used the string argument `stack_3v3` for this. Could you clarify the matter? Otherwise, the code is just awaiting some testing with the actual receiver.

DeflateAwning commented on firmware/Core/Src/gps/gps_internal_drivers.c:
```diff
@@ -0,0 +1,177 @@
+								  uint16_t rx_buf_len, const uint16_t rx_buf_max_size)
+{
+
+	// TX TO GPS
```
Please convert this file from tabs to spaces

DeflateAwning commented on firmware/Core/Src/rtos_tasks/rtos_tasks.c:
```diff
@@ -11,6 +11,9 @@
 #include "log/log.h"
 #include "eps_drivers/eps_commands.h"
+#include "gps/gps_types.h"
+#include "gps/gps.h"
```
Please revert autoformatted changes on this file

DeflateAwning commented on firmware/Core/Src/telecommands/telecommand_definitions.c:
```diff
@@ -25,6 +25,8 @@
 #include "timekeeping/timekeeping.h"
 #include "littlefs/littlefs_helper.h"
 #include "stm32/stm32_reboot_reason.h"
+#include "telecommands/eps_telecommands.h"
```
Please revert autoformatter changes on this file.

christopherkinyua replied:
Done

DeflateAwning commented on firmware/Core/Src/rtos_tasks/rtos_tasks.c:
```diff
@@ -197,26 +208,123 @@ void TASK_service_eps_watchdog(void *argument) {
 	}
 }
+
+void TASK_receive_gps_info(void *argument)
```
I'd recommend moving this task into a separate file (perhaps `gps_rtos_tasks.c`).

christopherkinyua replied:
Done

DeflateAwning reviewed (changes requested):
As requested, it looks good overall! There's a few things that should be fixed, and this branch should be rebased to main (or main merged into this branch), as it's starting to diverge a sizeable amount.

If there's a specific section you'd like me to check over, let me know! Obviously can't thoroughly review 2000 LOC several times.

NuclearTea commented on firmware/Core/Src/main.c:
```diff
@@ -289,6 +298,8 @@ int main(void)
 
   TASK_execute_telecommands_Handle = osThreadNew(TASK_execute_telecommands, NULL, &TASK_execute_telecommands_Attributes);
 
+  // TASK_receive_gps_Handle = osThreadNew(TASK_receive_gps_info, NULL, &TASK_receive_gps_attributes);
```
If this is not necessary, please remove it

NuclearTea commented on firmware/Core/Src/main.c:
```diff
@@ -251,7 +260,7 @@ int main(void)
   // Start the callback interrupts for the UART channels.
   UART_init_uart_handlers();
-  
+
```
remove this new line please

NuclearTea commented on firmware/Core/Src/main.c:
The vast majority of the changes in this file are spacing changes, can you please remove those changes.

NuclearTea commented on firmware/Core/Src/rtos_tasks/gps_rtos_tasks.c:
Isn't this apart of a different issue? If it is please separate this PR

NuclearTea commented on firmware/Core/Src/uart_handler/uart_handler.c:
Pretty sure this needs to be rebased lol

NuclearTea reviewed (commented):
Some comments. 

Please rebase. I think that a lot of the changes will be removed from that lol

NuclearTea commented:
This has been overhauled in another branch. This was the initial code written for GPS that was split into other PR's

DeflateAwning commented:
Which branch is this being continued in?

christopherkinyua commented:
> Which branch is this being continued in?

It is in the branch: chris/gps-ascii-parsers-i248. The PR for it is currently up as : GPS ASCII Parsers and Unit Tests #276
